Assertion failed 'regArgTab[argNum].slot == 2'

Seen in https://dev.azure.com/dnceng-public/public/_build/results?buildId=197177&view=logs&j=2f6a7d26-0d60-5ade-d191-981fe0847989&t=950a5a68-ae8a-5bb4-8cad-fb293bb99dc4&l=6041

(linux.arm64.Checked leg of runtime-extra-platforms)

```
  ILC: /__w/1/s/src/coreclr/jit/codegencommon.cpp:3957
  ILC: Assertion failed 'regArgTab[argNum].slot == 2' in 'System.Collections.Generic.GenericEqualityComparer`1[System.Numerics.Plane]:GetHashCode(System.Numerics.Plane):int:this' during 'Generate code' (IL size 24; hash 0x7039dd06; FullOpts)
  
  Aborted (core dumped)
/__w/1/s/artifacts/bin/coreclr/linux.arm64.Checked/build/Microsoft.NETCore.Native.targets(266,5): error MSB3073: The command ""/__w/1/s/artifacts/bin/coreclr/linux.arm64.Checked/x64/ilc/ilc" @"/__w/1/s/artifacts/obj/System.Numerics.Vectors.Tests/Release/net8.0/native/System.Numerics.Vectors.Tests.ilc.rsp"" exited with code 134. [/__w/1/s/src/libraries/System.Numerics.Vectors/tests/System.Numerics.Vectors.Tests.csproj]
```

Looks to be related to hardware intrinsics. Cc @tannergooding

For reference, the failure:
```
EXEC : error : Failed to load assembly 'SQLitePCLRaw.core' [/__w/1/s/src/libraries/System.Private.CoreLib/tests/IntrinsicsInSystemPrivatecoreLibAnalyzer.Tests/IntrinsicsInSystemPrivateCoreLib.Tests.csproj]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Failed to load assembly 'SQLitePCLRaw.core'
  Internal.TypeSystem.TypeSystemException+FileNotFoundException: Failed to load assembly 'SQLitePCLRaw.core'
     at Internal.TypeSystem.ThrowHelper.ThrowFileNotFoundException(ExceptionStringID id, String fileName) in /_/src/coreclr/tools/Common/TypeSystem/Common/ThrowHelper.cs:line 35
     at Internal.TypeSystem.ResolutionFailure.Throw() in /_/src/coreclr/tools/Common/TypeSystem/Common/ResolutionFailure.cs:line 105
     at Internal.TypeSystem.Ecma.EcmaModule.GetObject(EntityHandle handle, NotFoundBehavior notFoundBehavior) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs:line 432
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ResolveHandle(EntityHandle handle) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 61
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseTypeImpl(SignatureTypeCode typeCode) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 143
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseType(SignatureTypeCode typeCode) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 98
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseType() in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 309
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseFieldSignature() in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 462
     at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseFieldSignature(EmbeddedSignatureData[]& embeddedSigData) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs:line 471
     at Internal.TypeSystem.Ecma.EcmaField.InitializeFieldType() in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaField.cs:line 104
     at Internal.TypeSystem.MetadataFieldLayoutAlgorithm.ComputeInstanceLayout(DefType defType, InstanceLayoutKind layoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:line 37
     at Internal.TypeSystem.DefType.ComputeInstanceLayout(InstanceLayoutKind layoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/DefType.FieldLayout.cs:line 456
     at ILCompiler.CompilerTypeSystemContext.EnsureLoadableTypeUncached(TypeDesc type) in /_/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:line 317
     at ILCompiler.CompilerTypeSystemContext.EnsureLoadableType(TypeDesc type) in /_/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:line 74
     at ILCompiler.DependencyAnalysis.NodeFactory.CreateConstructedTypeNode(TypeDesc type) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:line 553
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
     at ILCompiler.DependencyAnalysis.ReflectionInvokeMapNode.AddSignatureDependency(DependencyList& dependencies, NodeFactory factory, TypeDesc type, String reason) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectionInvokeMapNode.cs:line 111
     at ILCompiler.DependencyAnalysis.ReflectedFieldNode.GetStaticDependencies(NodeFactory factory) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectedFieldNode.cs:line 96
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 182
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1 node) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 222
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 257
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 308
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 140
     at ILCompiler.Program.<Run>g__RunScanner|4_0(<>c__DisplayClass4_0&) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 468
     at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 448
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass221_0.<.ctor>b__0(ParseResult result) in /_/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 277
```

--- a/src/coreclr/jit/codegencommon.cpp
+++ b/src/coreclr/jit/codegencommon.cpp
@@ -2919,6 +2919,8 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
         bool writeThru;    // true if the argument gets homed to both stack and register
         bool processed;    // true after we've processed the argument (and it is in its final location)
         bool circular;     // true if this register participates in a circular dependency loop.
+        bool hfaConflict;  // arg is part of an HFA that will end up in the same register
+                           // but in a different slot (eg arg in s3 = v3.s[0], needs to end up in v3.s[3])
     } regArgTab[max(MAX_REG_ARG + 1, MAX_FLOAT_REG_ARG)] = {};
 
     unsigned   varNum;
@@ -3284,7 +3286,8 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
      * A circular dependency is a set of registers R1, R2, ..., Rn
      * such that R1->R2 (that is, R1 needs to be moved to R2), R2->R3, ..., Rn->R1 */
 
-    bool change = true;
+    bool change         = true;
+    bool hasHfaConflict = false;
     if (regArgMaskLive)
     {
         /* Possible circular dependencies still exist; the previous pass was not enough
@@ -3341,6 +3344,26 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
                     assert(regArgTab[argNum].slot <= (int)varDsc->lvHfaSlots());
                     assert(argNum > 0);
                     assert(regArgTab[argNum - 1].varNum == varNum);
+
+                    // If the field is passed in the same register as the destination,
+                    // but is in the wrong part of the register, mark it specially so later
+                    // we make sure to move it to the right spot before "freeing" the destination.
+                    //
+                    // If the arg is already in the right place we don't need to do anything special
+                    destRegNum = varDsc->GetRegNum();
+                    if (regNum == destRegNum)
+                    {
+                        const int slot = regArgTab[argNum].slot;
+                        if (slot != 1)
+                        {
+                            JITDUMP("HFA conflict; arg num %u needs to move from %s[%u] to %s[%u]\n", argNum,
+                                    getRegName(regNum), 0, getRegName(destRegNum), slot - 1);
+                            regArgTab[argNum].hfaConflict = true;
+
+                            // We'll need to do a special pass to resolve these
+                            hasHfaConflict = true;
+                        }
+                    }
                     regArgMaskLive &= ~genRegMask(regNum);
                     regArgTab[argNum].circular = false;
                     change                     = true;
@@ -3736,13 +3759,13 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
                 {
                     size = EA_4BYTE;
                 }
+                // HVA types...?
 
                 /* move the dest reg (begReg) in the extra reg */
 
                 assert(xtraReg != REG_NA);
 
                 regNumber begRegNum = genMapRegArgNumToRegNum(begReg, destMemType);
-
                 GetEmitter()->emitIns_Mov(insCopy, size, xtraReg, begRegNum, /* canSkip */ false);
 
                 regSet.verifyRegUsed(xtraReg);
@@ -3822,6 +3845,56 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
             }
         }
     }
+
+#if defined(TARGET_ARM64) && defined(FEATURE_SIMD)
+    // If we saw any hfa conflicts, handle those now
+    //
+    if (hasHfaConflict)
+    {
+        for (argNum = 0; argNum < argMax; argNum++)
+        {
+            if (!regArgTab[argNum].hfaConflict)
+            {
+                continue;
+            }
+
+            varNum                      = regArgTab[argNum].varNum;
+            varDsc                      = compiler->lvaGetDesc(varNum);
+            const regNumber destRegNum  = varDsc->GetRegNum();
+            const var_types regType     = regArgTab[argNum].type;
+            const unsigned  firstArgNum = argNum - (regArgTab[argNum].slot - 1);
+            const unsigned  lastArgNum  = firstArgNum + varDsc->lvHfaSlots() - 1;
+
+            assert(varDsc->lvIsHfa());
+            assert((argNum >= firstArgNum) && (argNum <= lastArgNum));
+            assert(destRegNum == genMapRegArgNumToRegNum(argNum, regType));
+
+            // Pass 0: move the conflicting part; Pass1: insert everthying else
+            //
+            for (int pass = 0; pass <= 1; pass++)
+            {
+                for (unsigned currentArgNum = firstArgNum; currentArgNum <= lastArgNum; currentArgNum++)
+                {
+                    const regNumber regNum = genMapRegArgNumToRegNum(currentArgNum, regType);
+                    bool            insertArg =
+                        ((pass == 0) && (currentArgNum == argNum)) || ((pass == 1) && (currentArgNum != argNum));
+
+                    if (insertArg)
+                    {
+                        assert(!regArgTab[currentArgNum].processed);
+
+                        // EA_4BYTE is probably wrong here (and below)
+                        // todo -- suppress self move
+                        GetEmitter()->emitIns_R_R_I_I(INS_mov, EA_4BYTE, destRegNum, regNum,
+                                                      regArgTab[currentArgNum].slot - 1, 0);
+                        regArgTab[currentArgNum].processed = true;
+                        regArgMaskLive &= ~genRegMask(regNum);
+                    }
+                }
+            }
+        }
+    }
+#endif // defined(TARGET_ARM64) && defined(FEATURE_SIMD)
 
     /* Finally take care of the remaining arguments that must be enregistered */
     while (regArgMaskLive)

--- a/src/coreclr/jit/codegencommon.cpp
+++ b/src/coreclr/jit/codegencommon.cpp
@@ -3340,6 +3340,7 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
                 {
                     // This must be a SIMD type that's fully enregistered, but is passed as an HFA.
                     // Each field will be inserted into the same destination register.
+                    //
                     assert(varTypeIsSIMD(varDsc));
                     assert(regArgTab[argNum].slot <= (int)varDsc->lvHfaSlots());
                     assert(argNum > 0);
@@ -3349,20 +3350,21 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
                     // but is in the wrong part of the register, mark it specially so later
                     // we make sure to move it to the right spot before "freeing" the destination.
                     //
-                    // If the arg is already in the right place we don't need to do anything special
                     destRegNum = varDsc->GetRegNum();
                     if (regNum == destRegNum)
                     {
+                        // We only get here if the HFA part is not already in the right slot in
+                        // the destination. That is, it is not slot-1.
+                        //
                         const int slot = regArgTab[argNum].slot;
-                        if (slot != 1)
-                        {
-                            JITDUMP("HFA conflict; arg num %u needs to move from %s[%u] to %s[%u]\n", argNum,
-                                    getRegName(regNum), 0, getRegName(destRegNum), slot - 1);
-                            regArgTab[argNum].hfaConflict = true;
+                        assert(slot != 1);
+                        JITDUMP("HFA conflict; arg num %u needs to move from %s[%u] to %s[%u]\n", argNum,
+                                getRegName(regNum), 0, getRegName(destRegNum), slot - 1);
+                        regArgTab[argNum].hfaConflict = true;
 
-                            // We'll need to do a special pass to resolve these
-                            hasHfaConflict = true;
-                        }
+                        // We'll need to do a special pass later to resolve these
+                        //
+                        hasHfaConflict = true;
                     }
                     regArgMaskLive &= ~genRegMask(regNum);
                     regArgTab[argNum].circular = false;
@@ -3847,10 +3849,29 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
     }
 
 #if defined(TARGET_ARM64) && defined(FEATURE_SIMD)
-    // If we saw any hfa conflicts, handle those now
+    // If we saw any hfa conflicts, handle those now.
     //
     if (hasHfaConflict)
     {
+        // Up above we noticed that there was at least one non-slot-1 HFA arg whose
+        // destination register was the same as the arg register.
+        //
+        // For example, say an HFA was passed as s0-s3 and the destination was v3.
+        // s3 is in the right register, but not in the right slot in the register.
+        //
+        // We handle this by first moving the conflicting part to the right slot
+        // in the destination (via pass 0 below), and then moving the remaining parts
+        // to their respective slots (via pass 1).
+        //
+        // Note the slot index in the register is one less than value of
+        // regArgTab[argNum].slot, so a slot-1 hfa arg goes into slot 0 of the destination).
+        //
+        // So for the above example, we'd first move the "slot-4" s3 (== v3.s[0]) to v3.s[3].
+        // Then we can insert s0 to v3.s[0]) and so on.
+        //
+        // We can exempt slot-1 cases as the conflicting part is already in the
+        // right slot, and code lower down correctly handles populating the remaining slots.
+        //
         for (argNum = 0; argNum < argMax; argNum++)
         {
             if (!regArgTab[argNum].hfaConflict)
@@ -3869,7 +3890,7 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
             assert((argNum >= firstArgNum) && (argNum <= lastArgNum));
             assert(destRegNum == genMapRegArgNumToRegNum(argNum, regType));
 
-            // Pass 0: move the conflicting part; Pass1: insert everthying else
+            // Pass 0: move the conflicting part; Pass1: insert everything else
             //
             for (int pass = 0; pass <= 1; pass++)
             {

--- a/src/libraries/tests.proj
+++ b/src/libraries/tests.proj
@@ -417,10 +417,6 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TestNativeAot)' == 'true' and '$(RunDisabledNativeAotTests)' != 'true'">
-    <!-- https://github.com/dotnet/runtime/issues/83167 -->
-    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Numerics.Vectors\tests\System.Numerics.Vectors.Tests.csproj"
-                       Condition="'$(TargetArchitecture)' == 'arm64'" />
-
     <!-- https://github.com/dotnet/runtime/issues/72908 -->
     <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Reflection.MetadataLoadContext\tests\System.Reflection.MetadataLoadContext.Tests.csproj" />
 

--- a/src/tests/JIT/Regression/JitBlue/Runtime_83167/Runtime_83167.cs
+++ b/src/tests/JIT/Regression/JitBlue/Runtime_83167/Runtime_83167.cs
@@ -1,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using Xunit;
+
+public class Runtime_83167
+{
+    [MethodImpl(MethodImplOptions.NoOptimization)]
+    [Fact]
+    public static int Problem()
+    {
+        Plane p = new Plane (new Vector3(2.0f, 3.0f, 4.0f), 1.0f);
+        int pH = p.GetHashCode();
+        EqualityComparer<Plane> c = EqualityComparer<Plane>.Default;
+        int cH = c.GetHashCode(p);
+        if (pH != cH)
+        {
+            Console.WriteLine($"Failed: {pH:X8} != {cH:X8}");
+            return 101;
+        }
+        return 100;
+    }
+}

--- a/src/tests/JIT/Regression/JitBlue/Runtime_83167/Runtime_83167.csproj
+++ b/src/tests/JIT/Regression/JitBlue/Runtime_83167/Runtime_83167.csproj
@@ -1,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <Optimize>True</Optimize>
+    <DebugType>None</DebugType>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="$(MSBuildProjectName).cs" />
+  </ItemGroup>
+</Project>

